src/commonMain/kotlin/exchange.dydx.abacus/processor/markets/CandlesProcessor.kt

```diff
@@ -2,12 +2,97 @@ package exchange.dydx.abacus.processor.markets
 
 import exchange.dydx.abacus.processor.base.BaseProcessor
 import exchange.dydx.abacus.protocols.ParserProtocol
+import exchange.dydx.abacus.state.internalstate.InternalMarketState
 import exchange.dydx.abacus.utils.mutable
 import exchange.dydx.abacus.utils.safeSet
+import indexer.codegen.IndexerCandleResponse
+import indexer.codegen.IndexerCandleResponseObject
+import kotlinx.datetime.Instant
 
-@Suppress("UNCHECKED_CAST")
-internal class CandlesProcessor(parser: ParserProtocol) : BaseProcessor(parser) {
-    private val itemProcessor = CandleProcessor(parser = parser)
+internal class CandlesProcessor(
+    parser: ParserProtocol,
+    private val itemProcessor: CandleProcessorProtocol = CandleProcessor(parser = parser),
+) : BaseProcessor(parser) {
+
+    fun processSubscribed(
+        existing: InternalMarketState,
+        resolution: String,
+        payload: IndexerCandleResponse?
+    ): InternalMarketState {
+        if (payload != null && (payload.candles?.size ?: 0) > 0) {
+            val candles = payload.candles?.reversed()?.mapNotNull {
+                itemProcessor.process(it)
+            }
+            val merged = merge(
+                parser = parser,
+                existing = existing.candles?.get(resolution),
+                incoming = candles,
+                timeField = {
+                    val startAt = it?.startedAtMilliseconds?.toLong()
+                    if (startAt != null) Instant.fromEpochMilliseconds(startAt) else null
+                },
+                ascending = true,
+            )
+            if (merged != null) {
+                val modified = existing.candles ?: mutableMapOf()
+                modified[resolution] = merged
+                existing.candles = modified
+            }
+        }
+        return existing
+    }
+
+    fun processBatchUpdate(
+        existing: InternalMarketState,
+        resolution: String,
+        payload: List<IndexerCandleResponseObject>?
+    ): InternalMarketState {
+        if (!payload.isNullOrEmpty()) {
+            val candles = payload.reversed().mapNotNull {
```

Comment:
why does the payload need to be reversed? Gut feeling is that the candle merge fn should be agnostic of the order and responsible for the sort instead of sorting payload first

Reply:
merge() assumes in inputs are sorted, which makes sense since this way it wouldn't need to call sort() every time.  It is more efficient with this assumption.